apps/emqx/src/emqx_crl_cache.erl

```diff
@@ -0,0 +1,320 @@
+
+start_link() ->
+    Config = gather_config(),
+    start_link(Config).
+
+start_link(Config = #{urls := _, refresh_interval := _, http_timeout := _}) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, Config, []).
+
+refresh(URL) ->
+    gen_server:cast(?MODULE, {refresh, URL}).
+
+evict(URL) ->
+    gen_server:cast(?MODULE, {evict, URL}).
+
+%% to pick up changes from the config
+-spec refresh_config() -> ok.
+refresh_config() ->
+    gen_server:cast(?MODULE, refresh_config).
+
+-spec refresh_config(map()) -> ok.
+refresh_config(Config) ->
+    gen_server:cast(?MODULE, {refresh_config, Config}).
+
+%%--------------------------------------------------------------------
+%% gen_server behaviour
+%%--------------------------------------------------------------------
+
+init(Config) ->
+    #{
+        urls := URLs,
+        refresh_interval := RefreshIntervalMS,
+        http_timeout := HTTPTimeoutMS
+    } = Config,
+    State = lists:foldl(
+        fun(URL, Acc) -> ensure_timer(URL, Acc, 0) end,
+        #state{
+            refresh_interval = RefreshIntervalMS,
+            http_timeout = HTTPTimeoutMS
+        },
+        URLs
+    ),
+    {ok, State}.
+
+handle_call(Call, _From, State) ->
+    {reply, {error, {bad_call, Call}}, State}.
+
+handle_cast({evict, URL}, State0 = #state{refresh_timers = RefreshTimers0}) ->
+    ssl_crl_cache:delete(URL),
+    MTimer = maps:get(URL, RefreshTimers0, undefined),
+    emqx_misc:cancel_timer(MTimer),
+    RefreshTimers = maps:without([URL], RefreshTimers0),
+    State = State0#state{refresh_timers = RefreshTimers},
+    ?tp(
+        crl_cache_evict,
+        #{url => URL}
+    ),
+    {noreply, State};
+handle_cast({refresh, URL}, State0) ->
+    case do_http_fetch_and_cache(URL, State0#state.http_timeout) of
+        {error, Error} ->
+            ?tp(crl_refresh_failure, #{error => Error, url => URL}),
+            ?SLOG(error, #{
+                msg => "failed_to_fetch_crl_response",
+                url => URL,
+                error => Error
+            }),
+            {noreply, ensure_timer(URL, State0, ?RETRY_TIMEOUT)};
+        {ok, _CRLs} ->
+            ?SLOG(debug, #{
+                msg => "fetched_crl_response",
+                url => URL
+            }),
+            {noreply, ensure_timer(URL, State0)}
+    end;
+handle_cast(refresh_config, State0) ->
+    #{
+        urls := URLs,
+        http_timeout := HTTPTimeoutMS,
+        refresh_interval := RefreshIntervalMS
+    } = gather_config(),
+    State = lists:foldl(
+        fun(URL, Acc) -> ensure_timer(URL, Acc, 0) end,
+        State0#state{
+            refresh_interval = RefreshIntervalMS,
+            http_timeout = HTTPTimeoutMS
+        },
+        URLs
+    ),
+    ?tp(crl_cache_refresh_config, #{
+        refresh_interval => RefreshIntervalMS,
+        http_timeout => HTTPTimeoutMS,
+        urls => URLs
+    }),
+    {noreply, State};
+handle_cast({refresh_config, Config}, State0) ->
+    case Config of
```

Comment:
nit: a little redundant, repeated the above cause

Reply:
True, I've refactored this bit.  Thanks!